File: Graphs/MinimumSpanningTree/Python/KruskalAdjacencyMatrix.py
class KruskalAdjacencyMatrix:
    def __init__(self, vertices):
        self.V = vertices
        self.adj = [[0] * vertices for _ in range(vertices)]
        self.parent = []
        self.rank = []

    def add_edge(self, u, v, w):
        self.adj[u][v] = w
        self.adj[v][u] = w

    def find(self, u):
        if u != self.parent[u]:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]

    def unite(self, u, v):
        pu, pv = self.find(u), self.find(v)
        if pu != pv:
            if self.rank[pu] < self.rank[pv]:
                pu, pv = pv, pu
            self.parent[pv] = pu
            if self.rank[pu] == self.rank[pv]:
                self.rank[pu] += 1

    def mst(self):
        edges = []
        for i in range(self.V):
            for j in range(i + 1, self.V):
                if self.adj[i][j] != 0:
                    edges.append((self.adj[i][j], i, j))
        edges.sort()

        self.parent = [i for i in range(self.V)]
        self.rank = [0] * self.V

        mst_edges = []
        for w, u, v in edges:
            if self.find(u) != self.find(v):
                self.unite(u, v)
                mst_edges.append((u, v, w))
        return mst_edges

File: Graphs/MinimumSpanningTree/CPP/KruskalAdjacencyMatrix.cpp
#include <vector>
#include <algorithm>
using namespace std;

class KruskalAdjacencyMatrix {
    int V;
    vector<vector<int>> adj;
    vector<int> parent, rank;

    int find(int u) {
        if (u != parent[u])
            parent[u] = find(parent[u]);
        return parent[u];
    }

    void unite(int u, int v) {
        int pu = find(u), pv = find(v);
        if (pu != pv) {
            if (rank[pu] < rank[pv]) swap(pu, pv);
            parent[pv] = pu;
            if (rank[pu] == rank[pv]) rank[pu]++;
        }
    }

public:
    KruskalAdjacencyMatrix(int vertices) : V(vertices) {
        adj.assign(V, vector<int>(V, 0));
    }

    void addEdge(int u, int v, int w) {
        adj[u][v] = w;
        adj[v][u] = w;
    }

    vector<tuple<int,int,int>> mst() {
        vector<tuple<int,int,int>> edges;
        for (int i = 0; i < V; i++)
            for (int j = i + 1; j < V; j++)
                if (adj[i][j] != 0)
                    edges.push_back({adj[i][j], i, j});

        sort(edges.begin(), edges.end());
        parent.resize(V);
        rank.assign(V, 0);
        for (int i = 0; i < V; i++) parent[i] = i;

        vector<tuple<int,int,int>> mstEdges;
        for (auto &[w, u, v] : edges) {
            if (find(u) != find(v)) {
                unite(u, v);
                mstEdges.push_back({u, v, w});
            }
        }
        return mstEdges;
    }
};
